### PR TITLE
new feature: tree-like conversation

### PR DESCRIPTION
**Pull request by littlemoonstones**
I have added a new feature: a tree-like conversation, but my code is a little bit messy and there is a bug that causes the loading animation to show immediately. (Fixed it in https://github.com/mckaywrigley/chatbot-ui/commit/fc3f3cf6ef2b92aaee8bdf61bb2d5c8c2b05e675)

https://user-images.githubusercontent.com/32943414/229262252-3533d692-9eb6-4081-994b-636d14ef69ac.mov

### CODE REVIEW

**Comment by mckaywrigley** [comment]
This is *really* cool. We'll definitely merge this idea in.

It needs just a bit of work around UI/UX and making sure it doesn't break any core things.

I'll bump this up on my todo list. Great start here.

**Comment by littlemoonstones** [comment]
Sorry, I have found the reason that I accidentally deleted some lines of code. Here's the link to the specific lines that were deleted: https://github.com/mckaywrigley/chatbot-ui/blob/aefec17525b664ba3e2d82512fa0196622925182/pages/index.tsx#L165-L199

File: pages/index.tsx [hunk 1]
```diff
@@ -528,12 +500,12 @@ const Home: React.FC<HomeProps> = ({
 
   // EFFECTS  --------------------------------------------
 
-  useEffect(() => {
-    if (currentMessage) {
-      handleSend(currentMessage);
-      setCurrentMessage(undefined);
-    }
-  }, [currentMessage]);
+  // useEffect(() => {
```
Comment (bcullman) [inline pages/index.tsx]: can this be removed?

File: utils/contexts/conversaionContext.tsx
Comment (bcullman) [inline utils/contexts/conversaionContext.tsx:1]: this filename is misspelled.

File: pages/index.tsx (hunk 1 above)
Comment (littlemoonstones) [reply pages/index.tsx]: Hi, I've addressed your comment and left the commented useEffect in the code. Please check the changes in commit c676f6fcf09b02831472d06e39d1c07a1d8f427d.

File: utils/contexts/conversaionContext.tsx
Comment (littlemoonstones) [reply utils/contexts/conversaionContext.tsx:1]: Thanks, I rename it in commit https://github.com/mckaywrigley/chatbot-ui/commit/fc9b66b9aa8d78483f30c6f7f037aa218d92c44f

**Comment by bcullman** [comment]
@littlemoonstones does this effect import/export data?

**Comment by littlemoonstones** [comment]
> @littlemoonstones does this effect import/export data?

I have tested it and found that it does not affect the import/export of data.

**Comment by RossHartmann** [comment]
> @littlemoonstones does this effect import/export data?

@bcullman I haven't yet tested on this branch but I suspect that after we make this modeling change, it would affect importing of any previous data formats since `Conversation` will no longer have a `messages` property -- while previous data formats *did* have a `messages` property.

I'll create a new data format to handle import/export of this new format in #377 then we can merge that into this branch.

UPDATE: I'm going to need to do the model changes on this branch since the changes greatly affect UI elements.

**Comment by RossHartmann** [comment]
@littlemoonstones I've created a [PR in your fork](https://github.com/littlemoonstones/chatbot-ui/pull/1) to merge into this branch. It updates the Conversation model and brings in the latest changes from `main`. Please let me know if you want to review anything.

**Comment by RossHartmann** [comment]
@bcullman Would you mind giving me more detail around why the Vercel bot failed to deploy to Preview? Looks like all unit tests are passing, but I can't see any details around the deploy failure.

**Comment by littlemoonstones** [comment]
> @bcullman Would you mind giving me more detail around why the Vercel bot failed to deploy to Preview? Looks like all unit tests are passing, but I can't see any details around the deploy failure.

IMG_03027C4A82D6-1

**Comment by RossHartmann** [comment]
Thank you, @littlemoonstones! Just created [another PR](https://github.com/littlemoonstones/chatbot-ui/pull/2) to fix this -- build and test both now pass.

**Comment by kaaboaye** [comment]
I think it might be better not to show those `<1/1>` and only show them after prompt was modified and there is something to do with those buttons.

If `<1/1>` were to be shown to all users every time it would create a lot of confusion especially that those buttons do not provide any value unless prompt was modified and there is `<2/2>`.

image

**Comment by bhaskoro-muthohar** [comment]
When can this feature be merged, guys? I need it so bad 🚀

**Comment by kaaboaye** [comment]
> When can this feature be merged, guys? I need it so bad 🚀

You can always run this brunch with tree-like directly instead waiting for merge.

**Comment by littlemoonstones** [comment]
> Sorry, I have found the reason that I accidentally deleted some lines of code. Here's the link to the specific lines that were deleted:
> 
> https://github.com/mckaywrigley/chatbot-ui/blob/aefec17525b664ba3e2d82512fa0196622925182/pages/index.tsx#L165-L199

I am trying to fix the message streaming in commit d8af3c7. Currently, it can display the streaming message successfully, but after the entire message shows up, the `selectedConversation` cannot be saved into localStorage. 



https://user-images.githubusercontent.com/32943414/229780827-9b9f8e9a-2906-4c4f-9d0d-4f0668d9899f.mov




The process is roughly like the diagram below:

```mermaid
sequenceDiagram
    participant handleSend
    participant selectedConversation
    participant chatNode_A
    participant messageStreaming
    participant saveConversation

    handleSend->>selectedConversation: call handleSend
    selectedConversation->>chatNode_A: add new chatNode(A)
    loop message streaming process
        messageStreaming->>chatNode_A: update content with text
    end
    chatNode_A->>saveConversation: call saveConversation
```
I guess the reason is related to asynchrony, but I cannot resolve the problem. Could anyone help me?

**Comment by littlemoonstones** [comment]
> I think it might be better not to show those `<1/1>` and only show them after prompt was modified and there is something to do with those buttons.
> 
> If `<1/1>` were to be shown to all users every time it would create a lot of confusion especially that those buttons do not provide any value unless prompt was modified and there is `<2/2>`.
> 
> image

OK, I have adjusted it in [cda77b5](https://github.com/mckaywrigley/chatbot-ui/pull/339/commits/cda77b58117a2d04924582c0cd0c101fe9c469f4).
image

**Comment by littlemoonstones** [comment]
> > Sorry, I have found the reason that I accidentally deleted some lines of code. Here's the link to the specific lines that were deleted:
> > https://github.com/mckaywrigley/chatbot-ui/blob/aefec17525b664ba3e2d82512fa0196622925182/pages/index.tsx#L165-L199
> 
> I am trying to fix the message streaming in commit [d8af3c7](https://github.com/mckaywrigley/chatbot-ui/commit/d8af3c7226c31f37d045e23c7a9a21e20bf7d247). Currently, it can display the streaming message successfully, but after the entire message shows up, the `selectedConversation` cannot be saved into localStorage.
> 
> The process is roughly like the diagram below:
> 
> 
> I guess the reason is related to asynchrony, but I cannot resolve the problem. Could anyone help me?

Fixed in fc3f3cf

**Comment by RossHartmann** [comment]
Anything we have left to do on this PR? @littlemoonstones

**Comment by littlemoonstones** [comment]
> Anything we have left to do on this PR? @littlemoonstones

The UI needs optimization, but I'm good at it. I wonder if it's necessary to optimize the UI for this PR?

**Comment by littlemoonstones** [comment]
> Anything we have left to do on this PR? @littlemoonstones

The UI needs optimization, but I'm not good at it. I wonder if it's necessary to optimize the UI for this PR?

**Comment by RossHartmann** [comment]
@mckaywrigley We've done a few UI optimizations. Any other changes you'd like to see on this PR before merging?

**Comment by moe001** [comment]
Is there any new progress on this job?

**Comment by RossHartmann** [comment]
> Is there any new progress on this job?

@moe001 I suspect this PR needs to go through some fairly massive changes due to the recent refactor in #494.

**Comment by littlemoonstones** [comment]
I have started to modify the code.

**Comment by littlemoonstones** [comment]
Currently, I commented on the function `handleDeleteMessage` in 7d990ab because I have no idea how to deal with deleting messages in a tree-like conversation.

**Pull request closed without merging**